libbeat/processors/rate_limit/config.go

```diff
@@ -0,0 +1,50 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package rate_limit
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/processors/rate_limit/algorithm"
+)
+
+// Config for rate limit processor.
+type Config struct {
+	Limit     algorithm.Rate         `config:"limit" validate:"required"`
+	Fields    []string               `config:"fields"`
+	Algorithm common.ConfigNamespace `config:"algorithm"`
+}
+
+func (c *Config) SetDefaults() error {
```

Comment:
This is what I had tried originally (see https://github.com/elastic/beats/pull/22883/commits/37bf29fee55e7324c8aab66ebe828e39c8c4963a). However, it doesn't work because `Algorithm` is of type `common.ConfigNamespace`, which enforces that only one algorithm can be configured at any given time. So first we configure the default, then if the user sets a different algorithm it doesn't have the effect of _overriding_ the default algorithm. Instead we get the error: `more than one namespace configured`.

That's why I ended up switching to the current implementation of setting defaults for any settings that are not already set by the user.

Reply:
Oh, interesting, ok then.